Update Block Templates Package for React 18

danielwrobert opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR fixes the TS error (post React 18 update) as follows:

1. Removed the unused `select as WPSelect` import since we're no longer using it directly.
2. Created a more generic `SelectType` type that's compatible with React 18.
3. Updated the `UseEvaluationContext` type to use this new `SelectType`.
4. Changed the `useSelect` hook to use `any` for the select parameter with appropriate ESLint comments.
5. Updated the `defaultUseEvaluationContext` function to match our new type definition.
6. Changed the `select` parameter type in `useEvaluationContext` from `typeof WPSelect` to `any` with appropriate ESLint comments.

These changes should bring the code up to compatibility with React 18 while maintaining the functionality of the code. The approach uses `any` in a controlled way with appropriate ESLint comments to acknowledge the type safety compromise, which is necessary given the comment about not being able to include proper type definitions.

This solution:

1. Fixes the immediate TypeScript error.
2. Maintains compatibility with React 18.
3. Preserves the existing functionality.
4. Follows the existing pattern of handling type issues in this file.

Sub-task of #55399.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Open the following files and confirm there are no TS errors introduced by this PR:
    - `packages/js/block-templates/src/utils/register-woo-block-type.ts`
    - `packages/js/product-editor/src/wp-hooks/hide-inventory-advanced-collapsible.tsx`
    - `packages/js/product-editor/src/utils/register-product-editor-block-type.ts`
2. Run tests from block-templates dir (`cd packages/js/block-templates && pnpm test:js`) and confirm all tests pass.
3. Run tests from product-editor dir (`cd packages/js/product-editor && pnpm test:js`) and confirm all tests pass.
4. Build the WooCommerce plugin via `pnpm --filter='@woocommerce/plugin-woocommerce' build` and confirm build succeeds without errors.

_Note: This PR only includes changes to types, there are no user-facing changes._

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

1. Ran tests from block-templates dir (`cd packages/js/block-templates && pnpm test:js`) and confirmed all tests pass.
2. Ran tests from product-editor dir (`cd packages/js/product-editor && pnpm test:js`) and confirmed all tests pass.
3. Built the WooCommerce plugin via `pnpm --filter='@woocommerce/plugin-woocommerce' build` and confirmed build succeeds without errors.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

Cleanup task: Fix TypeScript error in block templates package from React 18 upgrade.

</details>

danielwrobert committed:
Update register-woo-block-type.ts for React 18

These changes should resolve the TypeScript error while maintaining
the functionality of the code. The approach uses `any` in a
controlled way with appropriate ESLint comments to acknowledge the
type safety compromise, which is necessary given the comment about
not being able to include proper type definitions.

danielwrobert committed:
Update `select` param type in prod editor pkg

Changed the `select` parameter type in `useEvaluationContext`
from `typeof WPSelect` to `any` with appropriate ESLint comments.

Commit:
Add changefile(s) from automation for the following project(s): @woocommerce/product-editor, @woocommerce/block-templates

roykho commented on packages/js/block-templates/src/utils/register-woo-block-type.ts:
```diff
@@ -25,14 +29,13 @@ interface BlockRepresentation< T extends Record< string, object > > {
 	return {
-		getEvaluationContext: () => context,
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars
+		getEvaluationContext: ( select: any ) => context,
```
The argument (select) here isn't used, is there a reason we need that here?

danielwrobert replied:
Good catch, this is absolutely not needed!

Removed in bc29d98 and confirmed tests and builds are still passing, as expected.

roykho commented on packages/js/block-templates/src/utils/register-woo-block-type.ts:
```diff
@@ -55,8 +58,8 @@ function getEdit<
-			// @ts-expect-error TODO: react-18-upgrade
-			( select: typeof WPSelect ) => {
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+			( select: any ) => {
```
You've already declared the generic SelectType. Perhaps we can use that here so we don't need to use `any`?

danielwrobert replied:
Nice catch. Updated in b0964de.

roykho commented on packages/js/product-editor/src/utils/register-product-editor-block-type.ts:
```diff
@@ -20,7 +19,8 @@ export function useEvaluationContext( context: Record< string, unknown > ) {
 
-	const getEvaluationContext = ( select: typeof WPSelect ) => {
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	const getEvaluationContext = ( select: any ) => {
```
How come we don't also declare the generic `SelectType` here so we don't need to use `any`?

danielwrobert replied:
Good suggestion. I've added this in 9b818d9b05.

roykho reviewed (changes requested):
Thanks for working on this. I left some comments.

danielwrobert committed:
Remove unused arg from getEvaluationContext

The `select` arg is not used in the getEvaluationContext method. This
is removed here, per [this suggestion](https://github.com/woocommerce/woocommerce/pull/56341#discussion_r1993495077).

danielwrobert committed:
Use generic `SelectType` instead of `any`

Since we've already defined the generic `SelectType`, this should be
used instead of `any` in the `useSelect` callback arg.

danielwrobert committed:
Add `SelectType` generic to use instead of `any`

Replicate the `SelectType` generic we have defined in
`register-woo-block-type.ts` so we can use here instead of `any` as
well.

roykho commented on packages/js/block-templates/src/utils/register-woo-block-type.ts:
```diff
@@ -16,7 +16,11 @@ import { ComponentType } from 'react';
+import { useSelect } from '@wordpress/data';
+
+// Define a more generic type for the select function to avoid TypeScript errors
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
With the `any` type removed, we no longer need this comment.

danielwrobert replied:
Yeah, not sure why this was present to begin with. Good catch though, removed in 6b5aa8b3a5.

roykho commented on packages/js/product-editor/src/utils/register-product-editor-block-type.ts:
```diff
@@ -15,12 +14,17 @@ interface BlockRepresentation< T extends Record< string, object > > {
 }
 
+// Define a more generic type for the select function to avoid TypeScript errors
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
With the `any` type removed, we no longer need this comment.

danielwrobert replied:
Removed in 6b5aa8b.

roykho commented on packages/js/product-editor/src/utils/register-product-editor-block-type.ts:
```diff
@@ -15,12 +14,17 @@ interface BlockRepresentation< T extends Record< string, object > > {
 	const productId = useEntityId( 'postType', postType );
 
-	const getEvaluationContext = ( select: typeof WPSelect ) => {
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
With the `any` type removed, we no longer need this comment.

danielwrobert replied:
Right. Removed in 6b5aa8b3a5.

roykho reviewed (changes requested):
Thanks for the updates. I think we're almost there, just need to remove some comments.

danielwrobert committed:
Remove unneeded  comments

roykho commented on packages/js/product-editor/src/utils/register-product-editor-block-type.ts:
```diff
@@ -15,12 +14,15 @@ interface BlockRepresentation< T extends Record< string, object > > {
-	const getEvaluationContext = ( select: typeof WPSelect ) => {
+	const getEvaluationContext = ( select: SelectType ) => {
 		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
 		// @ts-ignore
```
Sorry I missed this part in the first two rounds. Is this something you want to also take care of in this PR? If you remove these ignore comments, you will reveal TS errors on the `select` because it does not have a type for the method `getEditedEntityRecord`

danielwrobert replied:
I wasn't intending to handle this bit, as my main focus here is the cleanup work in #55399 from the React 18 update. I'm not sure that this part would be related (unless I'm missing something).

That said, it seems a small enough change to where it should be fine to adjust while we're here.

danielwrobert replied:
Updated in 0ea6d67ded.

danielwrobert committed:
Add type assertion to select('core') return value

This adds a type assertion to properly type the return value
from `select('core')` by creating a coreStore variable with the
correct type signature for the `getEditedEntityRecord` method.

Additionally, it adds a type assertions for postType to explicitly
tell TypeScript that it should be treated as a string when passed
to both `useEntityId` and `getEditedEntityRecord`.

roykho committed:
Remove unneeded comments

roykho reviewed (approved):
Great work! Thanks for all the updates. I've pushed up a change to remove remaining unneeded comments on some dependencies because for some reason it didn't let me do inline code suggestion.

✅ I've tested `packages/js/block-templates/src/utils/register-woo-block-type.ts`
✅ I've tested `packages/js/product-editor/src/utils/register-product-editor-block-type.ts`
✅ I've tested `packages/js/product-editor/src/wp-hooks/hide-inventory-advanced-collapsible.tsx`
✅ I've ran the build script with no errors.
